native/cocos/audio/openharmony/AudioDecoder.cpp

```diff
@@ -0,0 +1,261 @@
+}
+
+bool AudioDecoder::init(const std::string &url, int sampleRate) {
+    _url = url;
+    _sampleRate = sampleRate;
+    return true;
+}
+
+bool AudioDecoder::start() {
+    auto oldTime = clockNow();
+    auto nowTime = oldTime;
+    bool ret;
+    do {
+        ret = decodeToPcm();
+        if (!ret) {
+            ALOGE("decodeToPcm (%s) failed!", _url.c_str());
+            break;
+        }
+
+        nowTime = clockNow();
+        ALOGD("Decoding (%s) to pcm data wasted %fms", _url.c_str(), intervalInMS(oldTime, nowTime));
+        oldTime = nowTime;
+
+        ret = resample();
+        if (!ret) {
+            ALOGE("resample (%s) failed!", _url.c_str());
+            break;
+        }
+
+        nowTime = clockNow();
+        ALOGD("Resampling (%s) wasted %fms", _url.c_str(), intervalInMS(oldTime, nowTime));
+        oldTime = nowTime;
+
+        ret = interleave();
+        if (!ret) {
+            ALOGE("interleave (%s) failed!", _url.c_str());
+            break;
+        }
+
+        nowTime = clockNow();
+        ALOGD("Interleave (%s) wasted %fms", _url.c_str(), intervalInMS(oldTime, nowTime));
+
+    } while (false);
+
+    ALOGV_IF(!ret, "%s returns false, decode (%s)", __FUNCTION__, _url.c_str());
+    return ret;
+}
+
+bool AudioDecoder::resample() {
+    if (_result.sampleRate == _sampleRate) {
+        ALOGI("No need to resample since the sample rate (%d) of the decoded pcm data is the same as the device output sample rate",
+              _sampleRate);
+        return true;
+    }
+
+    ALOGV("Resample: %d --> %d", _result.sampleRate, _sampleRate);
+
+    auto r = _result;
+    PcmBufferProvider provider;
+    provider.init(r.pcmBuffer->data(), r.numFrames, r.pcmBuffer->size() / r.numFrames);
+
+    const int outFrameRate = _sampleRate;
+    int outputChannels = 2;
+    size_t outputFrameSize = outputChannels * sizeof(int32_t);
+    size_t outputFrames = ((int64_t)r.numFrames * outFrameRate) / r.sampleRate;
+    size_t outputSize = outputFrames * outputFrameSize;
+    void *outputVAddr = malloc(outputSize);
+
+    auto resampler = AudioResampler::create(AUDIO_FORMAT_PCM_16_BIT, r.numChannels, outFrameRate,
+                                            AudioResampler::MED_QUALITY);
+    resampler->setSampleRate(r.sampleRate);
+    resampler->setVolume(AudioResampler::UNITY_GAIN_FLOAT, AudioResampler::UNITY_GAIN_FLOAT);
+
+    memset(outputVAddr, 0, outputSize);
+
+    ALOGV("resample() %zu output frames", outputFrames);
+
+    std::vector<int> Ovalues;
+
+    if (Ovalues.empty()) {
+        Ovalues.push_back(outputFrames);
+    }
+    for (size_t i = 0, j = 0; i < outputFrames;) {
+        size_t thisFrames = Ovalues[j++];
+        if (j >= Ovalues.size()) {
+            j = 0;
+        }
+        if (thisFrames == 0 || thisFrames > outputFrames - i) {
+            thisFrames = outputFrames - i;
+        }
+        int outFrames = resampler->resample((int *)outputVAddr + outputChannels * i, thisFrames,
```

Comment:
The channel count of resample is always the input channel count, here `outputChannels` is a constant 2.

The size of outputVAddr may be incorrect ?

Reply:
Android is implemented like this, I keep it here, leave it alone? It will be merged with android later

Reply:
I see